src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,209 @@
+            buffer.clear();
+            reader.join();
+        } catch (InterruptedException ie) {
+            throw new RuntimeException("Interrupted waiting for background thread to complete", ie);
+        } finally {
+            CloserUtil.close(underlying);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        throwOnCallingThread();
+        if (closeCalled) return false;
+        if (currentList != null && currentList.isEmpty()) {
+            // encountered end of stream
+            return false;
+        }
+        if (currentList == null || !currentBuffer.hasNext()) {
+            try {
+                currentList = buffer.take();
+                currentBuffer = currentList.iterator();
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Error reading from background thread", e);
+            }
+            // rethrow any exceptions raised on the background thread while we
+            // were blocking on the next record
+            throwOnCallingThread();
+        }
+        return currentBuffer.hasNext();
+    }
+
+    @Override
+    public T next() {
+        if (hasNext()) {
+            return (T) currentBuffer.next();
+        }
+        throw new NoSuchElementException("next");
+    }
+
+    /**
+     * Rethrows any exception encountered on the background
+     * thread to the caller 
+     */
+    private final void throwOnCallingThread() {
+        final Throwable t = this.ex.get();
+        if (t != null) {
+            if (t instanceof Error) {
+                throw (Error) t;
+            } else if (t instanceof RuntimeException) {
+                throw (RuntimeException) t;
+            } else {
+                throw new RuntimeException(t);
+            }
+        }
+    }
+
+    /**
+     * Reads the given iterator, passing back records to the calling thread in batches
+     */
+    private class ReaderRunnable implements Runnable {
+        public void run() {
+            boolean eosWritten = false;
+            try {
+                while (underlying.hasNext()) {
+                    final List<T> readAhead = new ArrayList<T>(bufferSize);
+                    for (int i = 0; i < bufferSize && underlying.hasNext(); i++) {
+                        readAhead.add(underlying.next());
+                    }
+                    buffer.put(readAhead);
+                    if (readAhead.isEmpty()) {
+                        eosWritten = true;
+                    }
+                }
+            } catch (InterruptedException ie) {
+                // log.debug("Thread interrupt received - closing on background thread.");
+            } catch (Throwable t) {
+                ex.set(t);
+                throw new RuntimeException(t);
+            } finally {
+                Thread.interrupted();
+                try {
+                    if (!eosWritten) {
+                        buffer.put(new ArrayList<T>());
```

Comment:
use `Collections.emptyList()` and add a comment that this empty list indicates the end of the underlying stream